docker/Makefile

```diff
@@ -0,0 +1,213 @@
+JUPYTER_LEDGER ?= $(PDO_LEDGER_URL)
+JUPYTER_MODE ?= build
+
+DOCKER_JUPYTER_COMMAND = "/project/pdo/tools/start_jupyter.sh"
+DOCKER_JUPYTER_COMMAND += -l $(JUPYTER_LEDGER)
+DOCKER_JUPYTER_COMMAND += --service-host $(JUPYTER_SERVICE_HOST)
+DOCKER_JUPYTER_COMMAND += -m $(JUPYTER_MODE)
+
+stop_jupyter :
+	- docker rm -f jupyter_container
+
+run_jupyter : build_contracts stop_jupyter
+	docker run $(DOCKER_RUN_ARGS) -it --name jupyter_container pdo_contracts:${CONTRACTS_VERSION} \
+		$(DOCKER_JUPYTER_COMMAND)
+
+# -----------------------------------------------------------------
+# We need a repository with the source for the branch we are going
+# to build. In theory this could just be a copy of the local source
+# tree but it might contain other artifacts. So instead we clone
+# the contracts with just a single branch in it. The storage and
+# performance requirements are relatively low.
+# -----------------------------------------------------------------
+contracts :
+	git clone --single-branch --branch $(CONTRACTS_BRANCH) --recurse-submodules $(CONTRACTS_REPO) contracts
+
+clean_contracts :
+	rm -rf contracts
+
+# -----------------------------------------------------------------
+# Testing uses docker-compose to build a set of images that will be configured
+# to run with localhost as the interface. in order to get a clean shutdown we
+# don't use a docker-compose specification for the base, ccf_base and services_base
+# containers. since we are rebuilding the contracts, we force recompilation of these
+# images from the rebuilt contracts. the corresponding images are built using docker
+# build instead of being included in the docker-compose.
+# -----------------------------------------------------------------
+
+# TEST_ARGS provides additional parameters for the compose
+# operations for running the test environment; generally
+# this would set in a make.loc
+TEST_COMPOSE_ARGS += -f test.yaml
+
+build_test : contracts models build_contracts versions
+
+test : clean_config clean_contracts build_test
+	- PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+	 	docker-compose $(TEST_COMPOSE_ARGS) down
+
+JUPYTER_COMPOSE_ARGS += -f test_jupyter.yaml
+
+test_jupyter : clean_config clean_contracts build_test
+	- PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+		docker-compose $(JUPYTER_COMPOSE_ARGS) up --abort-on-container-exit
+	PDO_VERSION=$(PDO_VERSION) CONTRACTS_VERSION=$(CONTRACTS_VERSION) \
+		docker-compose $(JUPYTER_COMPOSE_ARGS) down
+
+# -----------------------------------------------------------------
+# Cleaning is a bit interesting because the containers don't go away
+# unless they are told to very nicely. Until they go away they hold onto
+# referenced images. Clean make sure that everything is cleaned up
+# across the board.
```

Comment:
If you manually start, you can do that with `--rm` argument ...

Reply:
yeah... in this case i explicitly left it out. but i'm having second thoughts. the containers that are created are reasonably restarted (there is "client state" in them). HOWEVER... the real purpose of the docker directory is two-fold: 1) create images that others may find useful, 2) create a clean testing environment. the run_* commands aren't really consistent with either of those. so... i'm open to either approach.

Reply:
-rm added. still not sure if its the right thing but it will definitely make it easier to manage containers.